### PR TITLE
Switch to use NODE_OPS_CMD for decommission and bootstrap operation

### LINKED ISSUE #8471: Switch to use NODE_OPS_CMD for decommission operation
In commit 323f72e48a7d89d9edc40257a1f45c202accc146 (repair: Switch to
use NODE_OPS_CMD for replace operation), we switched replace operation
to use the new NODE_OPS_CMD infrastructure.

We should continue to switch decommission operation to use
NODE_OPS_CMD.

The benefits:

- A UUID is used to identify each node operation across the cluster.

- It is more reliable to detect pending node operations, to avoid
  multiple topology changes at the same time.

- The cluster reverts to a state before the decommission operation
  automatically in case of error. Without this patch, the node to be
  decommissioned will be stuck in decommission status forever until it
  is restarted and goes back to normal status.

- Allows users to pass a list of dead nodes to ignore for decommission
  explicitly.

- The LEAVING gossip status is not needed any more. This is one step
  closer to achieve gossip-less topology change.

- Allows us to trigger of off-strategy easily on the node receiving the
  ranges

### LINKED ISSUE #8472: Switch to use NODE_OPS_CMD for bootstrap operation
In commit 323f72e48a7d89d9edc40257a1f45c202accc146 (repair: Switch to
use NODE_OPS_CMD for replace operation), we switched replace operation
to use the new NODE_OPS_CMD infrastructure.

We should continue to switch bootstrap operation to use
NODE_OPS_CMD.

The benefits:

- It is more reliable to detect pending node operations, to avoid
  multiple topology changes at the same time, than using gossip status.

- The cluster reverts to a state before the bootstrap operation
  automatically in case of error much faster than gossip.

- Allows users to pass a list of dead nodes to ignore during bootstrap
  explicitly.

- The BOOTSTRAP gossip status is not needed any more. This is one step
  closer to achieve gossip-less topology change.

### PR DESCRIPTION
**Pull request by asias**
In commit 323f72e48a7d89d9edc40257a1f45c202accc146 (repair: Switch to
use NODE_OPS_CMD for replace operation), we switched replace operation
to use the new NODE_OPS_CMD infrastructure.

In this patch set, we continue the work to switch decommission and bootstrap 
operation to use NODE_OPS_CMD.

Fixes #8472
Fixes #8471

### CODE REVIEW

**Comment by asias** [comment]
"repair: Switch to use NODE_OPS_CMD for decommission operation" is needed to wire the off-strategy for decommission operation. 

"repair: Switch to use NODE_OPS_CMD for bootstrap operation", I want this patch to be part of 4.6 which is the first release to have NODE_OPS_CMD support, so we do not need to modify the data structure send over wire. 

@avikivity I know you are going to have vacation next week. It would be nice if we can speed up the view and settle it down before your vacation. 

With "Switch to use NODE_OPS_CMD for bootstrap operation", we are more strict about avoiding unsafe concurrent topology changes, dtest still bootstrap multiple nodes in parallel, we need to change dtest a lit to make the next gate passing. With my changes to dtest + this patch set, dtest next gating works. So don't panic if a test fails to start multiple nodes in parallel.

**Comment by asias** [comment]
On top of this series, I am going to send follow ups to

1) Wire up off-strategy for decommission
2) Share more code between different node operations for NODE_OPS_CMD

**Comment by bhalevy** [comment]
@tgrabiec @raphaelsc please review

File: service/storage_service.cc
Comment (avikivity) [inline service/storage_service.cc:1842]: What if the user did not enable repair_based_node_operations? Or if the cluster is not fully upgraded?

Comment (avikivity) [reply service/storage_service.cc:1842]: Ah, I see you handle non-upgraded clusters below.

File: service/storage_service.cc
Comment (avikivity) [inline service/storage_service.cc:1884]: Shouldn't you undo the operation for the nodes that did prepare successfully?

File: service/storage_service.cc
Comment (avikivity) [inline service/storage_service.cc:1905]: This (plus the code above) looks similar to the code in replace. Can it be shared?

File: service/storage_service.cc
Comment (avikivity) [inline service/storage_service.cc:1919]: I recommend you move the inside of this lambda to a named function, so you don't have to use `ss.` everywhere.

Comment (avikivity) [reply service/storage_service.cc:1919]: (I mean the seastar::async call, not parallel_for_each).

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1842]: Yes. Btw, I plan to move stream basded node operations back to removenode operations if repair_based_node_operations  is disabled. Currently, removenode only uses repair based node ops.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1884]: We do. we throw here and  do the revert in the catch clause.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1905]: I plan to do the refactor and share more code after this is merged.  I mentioned here too: https://github.com/scylladb/scylla/pull/8481#issuecomment-819268666.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1919]: I will consider it when I do the refactor after this series.

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:673]: please add comment about why this is done here only when RBNO is disabled.

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:1855]: Wouldn't a `std::unordered_set<gms::inet_address> ignore_nodes` be more efficient?

Comment (bhalevy) [reply service/storage_service.cc:1855]: Hmm, wait, `ignore_nodes` is empty at this time.
Worth a FIXME comment as you added below in `run_bootstrap_ops`

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:1881]: You can format the example in the warning message above using the actual node addresses listed in `nodes_down`.

File: service/storage_service.cc
Comment (bhalevy) [reply service/storage_service.cc:1884]: I'd add a comment here about this assumption here.

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:1887]: This doesn't seem like worth a thread.
Could easily be done with a `seastar::do_until` loop.

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:1901]: You can define another abort_source to be signaled when stopping heartbeat and just sleep_abortable the whole nr_seconds in one go, and be awakened immediately when heartbeat is stopped.

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:1847]: why do we need `heartbeat_updater_done` to be lw_shared?

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:1897]: I think it would be more interesting to warn about this per-node, not globally.
(Presuming the error will be related to specific nodes)

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:1934]: Is it OK to send an abort command to nodes you already succeeded sending decommission_done?
(I hope it is, given you send a uuid with it too).

If not, you can keep of list of those and ignore them too along with nodes_unknown_verb/down.

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:1987]: Shouldn't we check that for all repair-based node operations?
Or is it checked somewhere else for decommission / removenode?

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:2477]: minor: add op uuid to the message

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:2486]: ditto: add uuid to message

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:2489]: (preexisting) typo: s/node_ops_singal_abort/node_ops_signal_abort/

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:2567]: nit: add uuid to message

File: service/storage_service.cc
Comment (bhalevy) [inline service/storage_service.cc:2578]: nit: add uuid to message

**Review by bhalevy — COMMENTED** [review]
Looks good overall.  Please see individual comments for suggestions.

File: repair/repair.hh
Comment (asias) [reply repair/repair.hh:507]: I do not know if this is supposed. Since we need to modify idl files very  carefully than internal headers. Leaving it as a separate files makes more sense to me.  idl file has its own syntax, even if it looks similar to .hh file.

File: repair/repair.hh
Comment (asias) [reply repair/repair.hh:524]: fixed.

File: repair/repair.hh
Comment (asias) [reply repair/repair.hh:525]: fixed

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:657]: done

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1847]: because it is easier and safer to share between those two fibers.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1855]: To avoid big allocations if the list is large. We use list for the data we send over network. Keep it the same to avoid convert back and forth.

I added a TODO for dedecommission. We already have a TODO comment for run_bootstrap_ops.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1881]: The default formater will append the {} around it.  This example is clear enough in my view.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1884]: We already have a comment saying we need to revert in the catch clause.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1887]: This is on a supper slow path. do_until makes the code harder to write. What's the gain of converting go do_until?

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1901]: Actually, we can not abort here.  WE use the defer to wait for it. If we abort, we will terminate.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1934]: It is ok.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1987]: I am adding the wait logic here to make start multiple nodes in parallel to survive. We do not need it other operations.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:2489]: let's fix this as follow up.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:2477]: I will do it later as follow up.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:2486]: I will do it later as follow up.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:2578]: I will do it later as follow up.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:2567]: I will do it later as follow up.

File: service/storage_service.cc
Comment (asias) [reply service/storage_service.cc:1897]: Perhaps, I will leave it as it is for now. Since I plan to share more code as follow up.

File: repair/repair.hh
Comment (asias) [reply repair/repair.hh:531]: what is the real gain by adding noexcept here.

**Comment by asias** [comment]
Changes in v2:
- added missing space for node_ops_cmd_request constructor  
- added comments storage_service::bootstrap for updating pending ranges and gossip status
- added todo for ignore_nodes
-  drop the use of ss._abort_source
- logged exception when it fials to send heartbeat in storage_service::decommission

File: repair/repair.hh
Comment (avikivity) [reply repair/repair.hh:507]: (the plan is to use C++ reflection to create serializers/deserializers in C++ instead of idl-compiler.py)

https://en.cppreference.com/w/cpp/keyword/reflexpr

**Comment by avikivity** [comment]
I'll let @bhalevy review again.

File: service/storage_service.cc
Comment (bhalevy) [reply service/storage_service.cc:1892]: The gain is with conserving resources.  The do_until state is much lighter than a full blown async thread.
and I don't think the complexity in this loop justifies a thread.

**Review by bhalevy — APPROVED** [review]
I have no blocking comments.
We can fix the minor comments as a follow-up.

File: repair/repair.hh
Comment (asias) [reply repair/repair.hh:531]: The problem is do we really care about the exception safeness for this class? If not, why do we bother to type noexcept everywhere. Keep simple thing simple. IMHO, it is better not to use it by default.

**Comment by asias** [comment]
We need the dtest ccm/dtest fix: https://github.com/scylladb/scylla-ccm/pull/326 to pass the next-gating.

File: repair/repair.hh
Comment (bhalevy) [reply repair/repair.hh:531]: what's complicated with adding noexcept where it is obvious?
similarly, why do we care about `const`? we could do without it too...

Comment (asias) [reply repair/repair.hh:531]: Sometimes maybe it is obvious and sometimes it is not. We use use a c++ feature not because it is obvious. Adding noexcept here will cause the whole scylla process to terminate if we the function actually throws.  In no ways, such user operations should cause a scylla process to terminate.

Comment (bhalevy) [reply repair/repair.hh:531]: > Adding noexcept here will cause the whole scylla process to terminate if we the function actually throws.
But it doesn't.
Maybe this case is too high level, but it's very important for lower level api's to be marked noexcept (when they are) so that their users can rely on that.  The boost library, for example, rarely uses `noexcept`, unlike the standard c++ library and it got us into trouble more than once when, for example, we require lambda functions to be nothrow move constructible and they capture a boost container which move constructor is not marked noexcept.

**Comment by asias** [comment]
> We need the ccm/dtest fix: [scylladb/scylla-ccm#326](https://github.com/scylladb/scylla-ccm/pull/326) to pass the next-gating.

ccm/dtest fix is not needed. 

@avikivity Ping for merge.

**Comment by asias** [comment]
@avikivity Ping for merge ^2.

File: idl/partition_checksum.idl.hh
Comment (avikivity) [inline idl/partition_checksum.idl.hh:142]: Please document for which verbs this is expected to have a value.

These multiplexed verbs are hard to understand.

File: idl/partition_checksum.idl.hh
Comment (avikivity) [inline idl/partition_checksum.idl.hh:137]: So, different verbs will have different fields set in this structure?

Please add documentation.

File: idl/partition_checksum.idl.hh
Comment (asias) [reply idl/partition_checksum.idl.hh:142]: Ok. Done.

File: idl/partition_checksum.idl.hh
Comment (asias) [reply idl/partition_checksum.idl.hh:137]: Yes, documentation is added.

**Comment by asias** [comment]
Changes in v3:
- Added more docs for node_ops_cmd_request and node_ops_cmd_response.

**Comment by bhalevy** [comment]
@avikivity please re-review

**Comment by bhalevy** [comment]
ping @avikivity

**Comment by asias** [comment]
> Isn't this prone to the safety problem brought up during review for the replace operation in [#8330 (review)](https://github.com/scylladb/scylla/pull/8330#pullrequestreview-625093341) ?
> 
> For replace we concluded there was no problem if existing node reverts temporarily to the old topology on restart because the quorum in the old replica set must also be a quorum in the new replica set. During replace, in the replica set of a given partition only the node being replaced changes, and it is either permanently down (replace-with-a-different-ip) so doesn't participate in quorum, or it is the new node (replace-with-same-ip) in which case the replica set does not change.
> 
> But for bootstrap, replica set can change from (A, B, C) to (A, B, D). If some node reverts to (A, B, C), it can ACK an operation using a quorum which is not a quorum in the new replica set, e.g (B, C).
> 
> \cc @avikivity @kostja

Are you referring to the existing node restarts during bootstrap? Bootstrap does not have the issues like we have in replace. 

If an existing node restarts after all nodes ACK the bootstrap_done, it will learn the new topology from gossip during restart. All nodes will use new topology. 

If an existing node restarts before all nodes ACK the bootstrap_done, bootstrap will fail. The whole cluster will revert to use old topology.  When the existing node restarts, it will not learn the new topology. It does not matter if the node ack cql request that only satisfy the old replica set.

**Comment by tgrabiec** [comment]
> > Isn't this prone to the safety problem brought up during review for the replace operation in [#8330 (review)](https://github.com/scylladb/scylla/pull/8330#pullrequestreview-625093341) ?
> > For replace we concluded there was no problem if existing node reverts temporarily to the old topology on restart because the quorum in the old replica set must also be a quorum in the new replica set. During replace, in the replica set of a given partition only the node being replaced changes, and it is either permanently down (replace-with-a-different-ip) so doesn't participate in quorum, or it is the new node (replace-with-same-ip) in which case the replica set does not change.
> > But for bootstrap, replica set can change from (A, B, C) to (A, B, D). If some node reverts to (A, B, C), it can ACK an operation using a quorum which is not a quorum in the new replica set, e.g (B, C).
> > \cc @avikivity @kostja
> 
> Are you referring to the existing node restarts during bootstrap? 

Yes

> Bootstrap does not have the issues like we have in replace.
> 
> If an existing node restarts after all nodes ACK the bootstrap_done, it will learn the new topology from gossip during restart. All nodes will use new topology.

But during restart we don't wait for `ring_delay`. A node could boot before the gossip change propagates to the new node. 

It's also not partition tolerant, if there is a network partition which prevents gossip to propagate, the restarted node will serve requests using old topology.

**Comment by tgrabiec** [comment]
On Tue, May 11, 2021 at 2:29 AM Asias He ***@***.***> wrote:

> Isn't this prone to the safety problem brought up during review for the
> replace operation in #8330 (review)
> <https://github.com/scylladb/scylla/pull/8330#pullrequestreview-625093341>
> ?
> For replace we concluded there was no problem if existing node reverts
> temporarily to the old topology on restart because the quorum in the old
> replica set must also be a quorum in the new replica set. During replace,
> in the replica set of a given partition only the node being replaced
> changes, and it is either permanently down (replace-with-a-different-ip) so
> doesn't participate in quorum, or it is the new node (replace-with-same-ip)
> in which case the replica set does not change.
> But for bootstrap, replica set can change from (A, B, C) to (A, B, D). If
> some node reverts to (A, B, C), it can ACK an operation using a quorum
> which is not a quorum in the new replica set, e.g (B, C).
> \cc @avikivity <https://github.com/avikivity> @kostja
> <https://github.com/kostja>
>
> Are you referring to the existing node restarts during bootstrap?
>
> Yes
>
> Bootstrap does not have the issues like we have in replace.
> If an existing node restarts after all nodes ACK the bootstrap_done, it
> will learn the new topology from gossip during restart. All nodes will use
> new topology.
>
> But during restart we don't wait for ring_delay. A node could boot before
> the gossip change propagates to the new node.
>
> During restart, a node waits for gossip to settle.
>

It's a short wait though. Is it enough to guarantee propagation? I thought
we have much longer ring_delay wait during topology changes because
the shorter wait is not safe enough in large clusters.

It's true that ring_delay doesn't guarantee that either, but at least it's
what is considered relatively safe under normal conditions.

It's also not partition tolerant, if there is a network partition which
> prevents gossip to propagate, the restarted node will serve requests using
> old topology.
>
> Isn't this an existing issue. This can happen before this patch too. So
> far, we still use the NORMAL status.
>

But before this patch, we set the BOOTSTRAPPING state in gossip before we
start streaming. If that state doesn't reach the node, streaming will also
probably fail. After streaming, when we jump to NORMAL, if that doesn't
reach the partitioned node it will still use both old and new topology when
it learns about BOOTSTRAPPING from any neighbor, so should be safer.

If many nodes restart at the same time, the old method was also not safe
because BOOTSTRAPPING is not stored in local system tables and has to reach
us via gossip. That's https://github.com/scylladb/scylla/issues/6088

So the old method was also prone to those problems under some scenarios,
but the new method is less safe. I want us to be aware of that.

It could be made safe if we persist the state.

**Pull request merged**
